File: MATTHEW/leetcode/AIcoding/2.Gemini-2.0/cpp_solutions/514-freedom-trail.cpp
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    int findRotateSteps(string ring, string key) {
        int n = ring.size();
        int m = key.size();

        vector<vector<int>> pos(26);
        for (int i = 0; i < n; ++i) {
            pos[ring[i] - 'a'].push_back(i);
        }

        vector<vector<int>> dp(m + 1, vector<int>(n, 1e9));
        for (int i = 0; i < n; ++i) {
            dp[0][i] = 0;
        }

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (dp[i][j] != 1e9) {
                    for (int k : pos[key[i] - 'a']) {
                        int diff = abs(j - k);
                        int cost = min(diff, n - diff);
                        dp[i + 1][k] = min(dp[i + 1][k], dp[i][j] + cost + 1);
                    }
                }
            }
        }

        int ans = 1e9;
        for (int i = 0; i < n; ++i) {
            ans = min(ans, dp[m][i]);
        }

        return ans;
    }
};

File: MATTHEW/leetcode/AIcoding/2.Gemini-2.0/rust_solutions/514-freedom-trail.rs
impl Solution {
    pub fn find_rotate_steps(ring: String, key: String) -> i32 {
        let ring_chars: Vec<char> = ring.chars().collect();
        let key_chars: Vec<char> = key.chars().collect();
        let m = ring_chars.len();
        let n = key_chars.len();
        let mut pos: Vec<Vec<usize>> = vec![vec![]; 26];
        for i in 0..m {
            pos[(ring_chars[i] as u8 - b'a') as usize].push(i);
        }
        let mut dp: Vec<Vec<i32>> = vec![vec![i32::MAX; m]; n + 1];
        for &p in &pos[(key_chars[0] as u8 - b'a') as usize] {
            dp[0][p] = std::cmp::min(p as i32, (m - p) as i32) + 1;
        }
        for i in 1..n {
            for &j in &pos[(key_chars[i] as u8 - b'a') as usize] {
                for &k in &pos[(key_chars[i - 1] as u8 - b'a') as usize] {
                    let diff = (j as i32 - k as i32).abs();
                    let step = std::cmp::min(diff, m as i32 - diff);
                    if dp[i - 1][k] != i32::MAX {
                        dp[i][j] = std::cmp::min(dp[i][j], dp[i - 1][k] + step + 1);
                    }
                }
            }
        }
        let mut ans = i32::MAX;
        for &p in &pos[(key_chars[n - 1] as u8 - b'a') as usize] {
            ans = std::cmp::min(ans, dp[n - 1][p]);
        }
        ans
    }
}